ext/reflection/php_reflection.c

```diff
@@ -5715,6 +5715,108 @@ ZEND_METHOD(reflection_property, hasType)
 }
 /* }}} */
 
+/* {{{ proto public bool ReflectionProperty::hasDefaultValue()
+   Returns whether property has a default value */
+ZEND_METHOD(reflection_property, hasDefaultValue)
+{
+	reflection_object *intern;
+	property_reference *ref;
+	zend_property_info *prop_info;
+	zval *prop, prop_copy;
+	zend_class_entry *ce;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	GET_REFLECTION_OBJECT_PTR(ref);
+
+	prop_info = ref->prop;
+
+	if (prop_info == NULL) {
+		RETURN_FALSE;
+	}
+
+	ce = prop_info->ce;
+
+	if ((prop_info->flags & ZEND_ACC_STATIC) != 0) {
+		prop = &ce->default_static_members_table[prop_info->offset];
+		ZVAL_DEINDIRECT(prop);
+	} else {
+		prop = &ce->default_properties_table[OBJ_PROP_TO_NUM(prop_info->offset)];
+	}
+
+	if (Z_ISUNDEF_P(prop)) {
+		RETURN_FALSE;
+	}
+
+	/* copy: enforce read only access */
+	ZVAL_DEREF(prop);
+	ZVAL_COPY_OR_DUP(&prop_copy, prop);
+
+	/* this is necessary to make it able to work with default array
+	* properties, returned to user */
+	if (Z_TYPE(prop_copy) == IS_CONSTANT_AST) {
+		if (UNEXPECTED(zval_update_constant_ex(&prop_copy, ce) != SUCCESS)) {
+			RETURN_THROWS();
+		}
+	}
+
+	RETURN_TRUE;
+}
+/* }}} */
+
+/* {{{ proto public mixed ReflectionProperty::getDefaultValue()
+   Returns the default value of a property */
+ZEND_METHOD(reflection_property, getDefaultValue)
+{
+	reflection_object *intern;
+	property_reference *ref;
+	zend_property_info *prop_info;
+	zval *prop, prop_copy;
+	zend_class_entry *ce;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	GET_REFLECTION_OBJECT_PTR(ref);
+
+	prop_info = ref->prop;
+
+	if (prop_info == NULL) {
+		return; // throw exception?
+	}
+
+	ce = prop_info->ce;
+
+	if ((prop_info->flags & ZEND_ACC_STATIC) != 0) {
+		prop = &ce->default_static_members_table[prop_info->offset];
+		ZVAL_DEINDIRECT(prop);
+	} else {
+		prop = &ce->default_properties_table[OBJ_PROP_TO_NUM(prop_info->offset)];
+	}
+
+	if (Z_ISUNDEF_P(prop)) {
+		return;
+	}
+
+	/* copy: enforce read only access */
+	ZVAL_DEREF(prop);
+	ZVAL_COPY_OR_DUP(&prop_copy, prop);
+
+	/* this is necessary to make it able to work with default array
+	* properties, returned to user */
+	if (Z_TYPE(prop_copy) == IS_CONSTANT_AST) {
+		if (UNEXPECTED(zval_update_constant_ex(&prop_copy, ce) != SUCCESS)) {
+			RETURN_THROWS();
+		}
+	}
+
+	RETURN_ZVAL(prop, 1, 0);
```

Comment:
This should be returning `prop_copy`. Though it's better to instead of `ZVAL_COPY_OR_DUP(return_value, prop)` above. You may want to add a test that needs to evaluate a constant expression, to verify that the right thing is being returned in that case.

(In PHP 7 it's best to avoid use of RETURN_ZVAL. Nobody knows what those two booleans mean.)

Reply:
Oh that is what CONSTANT_AST is :-) Added the test and your change was indeed necessary.